File: 11_动态规划/lcs最长公共子序列问题/最长公共子序列/位运算加速LCS.ts
/* eslint-disable no-inner-declarations */
// 位运算加速LCS(最长公共子序列), 这也是目前跑的最快的版本
// https://wenku.baidu.com/view/ed99e4f77c1cfad6195fa776.html?_wkts_=1688318339264
// https://atcoder.jp/contests/dp/tasks/dp_f
// https://www.cnblogs.com/-Wallace-/p/bit-lcs.html
// https://rsk0315.hatenablog.com/entry/2022/12/30/180216
// https://github.com/hqztrue/LeetCodeSolutions/blob/master/0501-0600/516.%20Longest%20Palindromic%20Subsequence.pdf
// https://github.com/hqztrue/LeetCodeSolutions/blob/master/1101-1200/1143.%20Longest%20Common%20Subsequence%200ms.cpp
// https://loj.ac/s?problemDisplayId=6564&status=Accepted

/**
 * 位运算加速LCS(最长公共子序列).
 * @complex `O(nm/32)`.1e5x1e5 => 1.5s.
 */
function LCSBit<T>(arr1: ArrayLike<T>, arr2: ArrayLike<T>): number {
  const n = arr1.length
  const m = arr2.length
  const copy1 = new Uint32Array(n)
  const copy2 = new Uint32Array(m)

  const id = new Map<T, number>()
  for (let i = 0; i < n; ++i) {
    const item = arr1[i]
    const curId = id.get(item)
    if (curId !== void 0) {
      copy1[i] = curId
    } else {
      copy1[i] = id.size
      id.set(item, id.size)
    }
  }
  for (let i = 0; i < m; ++i) {
    const item = arr2[i]
    const curId = id.get(item)
    if (curId !== void 0) {
      copy2[i] = curId
    } else {
      copy2[i] = id.size
      id.set(item, id.size)
    }
  }

  const f = Array<_BS>(id.size)
  for (let i = 0; i < id.size; ++i) f[i] = new _BS(n)
  for (let i = 0; i < n; ++i) f[copy1[i]].set(i)
  const dp = new _BS(n)
  for (let i = 0; i < m; ++i) dp.run(f[copy2[i]])
  return dp.count()
}

class _BS {
  private static _onesCount32(uint32: number): number {
    uint32 -= (uint32 >>> 1) & 0x55555555
    uint32 = (uint32 & 0x33333333) + ((uint32 >>> 2) & 0x33333333)
    return (((uint32 + (uint32 >>> 4)) & 0x0f0f0f0f) * 0x01010101) >>> 24
  }

  private readonly _data: Uint32Array

  constructor(n: number) {
    this._data = new Uint32Array(1 + ((n / 31) | 0))
  }

  set(i: number): void {
    this._data[(i / 31) | 0] |= 1 << i % 31
  }

  count(): number {
    let res = 0
    for (let i = 0; i < this._data.length; ++i) {
      res += _BS._onesCount32(this._data[i])
    }
    return res
  }

  run(o: _BS): void {
    let c = 1
    for (let i = 0; i < this._data.length; ++i) {
      const v = this._data[i]
      let x = v
      const y = v | o._data[i]
      x += x + c + (~y & 0x7fffffff) // ((1 << 31) >>> 0) - 1
      this._data[i] = x & y
      c = x >>> 31
    }
  }
}

export { LCSBit }

if (require.main === module) {
  // https://leetcode.cn/problems/longest-common-subsequence/
  function longestCommonSubsequence(text1: string, text2: string): number {
    return LCSBit(text1, text2)
  }

  const n1 = 1e5
  const n2 = 1e5
  const arr1 = Array.from({ length: n1 }, () => ~~(Math.random() * n2))
  const arr2 = Array.from({ length: n1 }, () => ~~(Math.random() * n2))
  console.time('LCSBit')
  console.log(LCSBit(arr1, arr2))
  console.timeEnd('LCSBit')
}

File: 11_动态规划/lcs最长公共子序列问题/最长公共子序列/位运算加速LCS.go
// lcsbit/bitlcs
// https://atcoder.jp/contests/dp/submissions/34604402
// https://loj.ac/s/1633431

package main

import (
	"bufio"
	"fmt"
	"math/bits"
	"os"
	"time"
)

func main() {
	test()
}

func test() {
	// 10000*10000
	nums1 := make([]int, int(1e5))
	nums2 := make([]int, int(1e5))
	for i := range nums1 {
		nums1[i] = i
		nums2[i] = i
	}
	time1 := time.Now()
	fmt.Println(LCSBit(nums1, nums2))
	fmt.Println(time.Since(time1))
}

func demo() {
	in := bufio.NewReader(os.Stdin)
	out := bufio.NewWriter(os.Stdout)
	defer out.Flush()

	var s1, s2 string
	fmt.Fscan(in, &s1, &s2)
	fmt.Fprintln(out, LCSBit2(s1, s2))
}

func LCSBit2(s1, s2 string) int {
	ords1, ords2 := make([]int, len(s1)), make([]int, len(s2))
	for i, v := range s1 {
		ords1[i] = int(v)
	}
	for i, v := range s2 {
		ords2[i] = int(v)
	}
	return LCSBit(ords1, ords2)
}

// 位运算加速LCS(最长公共子序列).
func LCSBit(arr1 []int, arr2 []int) int {
	arr1, arr2 = append(arr1[:0:0], arr1...), append(arr2[:0:0], arr2...)
	id := make(map[int]int)
	for i, v := range arr1 {
		if _, ok := id[v]; !ok {
			id[v] = len(id)
		}
		arr1[i] = id[v]
	}
	for i, v := range arr2 {
		if _, ok := id[v]; !ok {
			id[v] = len(id)
		}
		arr2[i] = id[v]
	}

	n := len(arr1)
	size := len(id)
	f := make([]*BS, size)
	for i := range f {
		f[i] = NewBS(n)
	}
	for i, v := range arr1 {
		f[v].Set(i)
	}

	dp := NewBS(n)
	for _, v := range arr2 {
		dp.Run(f[v])
	}
	return dp.Count()
}

type BS struct {
	data []uint64
}

func NewBS(n int) *BS {
	return &BS{
		data: make([]uint64, 1+(n/63)),
	}
}

func (bs *BS) Set(i int) {
	bs.data[i/63] |= 1 << (i % 63)
}

func (bs *BS) Count() int {
	res := 0
	for _, v := range bs.data {
		res += bits.OnesCount64(v)
	}
	return res
}

func (bs *BS) Run(o *BS) {
	c := uint64(1)
	for i, v := range bs.data {
		x, y := v, v|o.data[i]
		x += x + c + (^y & (1<<63 - 1))
		bs.data[i] = x & y
		c = x >> 63
	}
}
